odoo/websocket.py

```diff
@@ -0,0 +1,959 @@
+            # is set to ensure the connection will be closed even if
+            # the other end does not close the socket properly.
+            self._socket.settimeout(1)
+            remaining_data = self._socket.recv(4096)
+            while remaining_data:
+                remaining_data = self._socket.recv(4096)
+        self._socket.close()
+        self.state = ConnectionState.CLOSED
+        self._trigger_lifecycle_event(LifecycleEvent.CLOSE)
+
+    def _handle_control_frame(self, frame):
+        if frame.opcode is Opcode.PING:
+            self._enqueue_pong_frame(frame.payload)
+        elif frame.opcode is Opcode.CLOSE:
+            self.state = ConnectionState.CLOSING
+            code, reason = CloseCode.CLEAN, None
+            if len(frame.payload) >= 2:
+                code = struct.unpack('!H', frame.payload[:2])[0]
+                reason = frame.payload[2:].decode('utf-8')
+            elif frame.payload:
+                raise ProtocolError("Malformed closing frame")
+            if not self._close_sent:
+                self._enqueue_close_frame(code, reason)
+            else:
+                self._terminate()
+
+    def _handle_transport_error(self, exc):
+        """
+            Find out which close code should be sent according to given
+            exception and call `self.disconnect` in order to close the
+            connection cleanly.
+        """
+        if self.state is ConnectionState.CLOSED:
+            return
+        code, reason = CloseCode.SERVER_ERROR, str(exc)
+        if isinstance(exc, (ConnectionClosed, OSError)):
+            code = CloseCode.ABNORMAL_CLOSURE
+        elif isinstance(exc, (ProtocolError, InvalidCloseCodeException)):
+            code = CloseCode.PROTOCOL_ERROR
+        elif isinstance(exc, UnicodeDecodeError):
+            code = CloseCode.INCONSISTENT_DATA
+        elif isinstance(exc, PayloadTooLargeException):
+            code = CloseCode.MESSAGE_TOO_BIG
+        elif isinstance(exc, RateLimitExceededException):
+            code = CloseCode.TRY_LATER
+        if code is CloseCode.SERVER_ERROR:
+            _logger.error(exc, exc_info=True)
+        self.disconnect(code, reason)
+
+    def _limit_rate(self):
+        """
+            This method is a simple rate limiter designed not to allow
+            more than one request by `RL_DELAY` seconds. `RL_BURST` specify
+            how many requests can be made in excess of the given rate at the
+            begining. When requests are received too fast, raises the
+            `RateLimitExceededException`.
+        """
+        now = time.time()
+        if len(self._inncoming_frame_timestamps) == type(self).RL_BURST:
+            ellapsed_time = now - self._inncoming_frame_timestamps[0]
+            if ellapsed_time < type(self).RL_DELAY * type(self).RL_BURST:
+                raise RateLimitExceededException()
+        self._inncoming_frame_timestamps.append(now)
```

Comment:
```suggestion
        now = time.time()
        if len(self._incoming_frame_timestamps) == type(self).RL_BURST:
            elapsed_time = now - self._incoming_frame_timestamps[0]
            if elapsed_time < type(self).RL_DELAY * type(self).RL_BURST:
                raise RateLimitExceededException()
        self._incoming_frame_timestamps.append(now)
```

Beware the `append` on a maxlen-less deque. See my above comment.

Reply:
Fixed, test added :+1: